packages/office-ui-fabric-react/src/utilities/selection/Selection.ts

```diff
@@ -135,37 +135,43 @@ export class Selection implements ISelection {
       }
     }
 
-    if (shouldClear) {
+    if (shouldClear || items.length === 0) {
       this.setAllSelected(false);
     }
 
     // Check the exemption list for discrepencies.
     const newExemptedIndicies: { [key: string]: boolean } = {};
+    let newExemptedCount = 0;
 
     for (const indexProperty in this._exemptedIndices) {
-      if (this._exemptedIndices.hasOwnProperty(indexProperty)) {
+      if (Object.prototype.hasOwnProperty.call(this._exemptedIndices, indexProperty)) {
         const index = Number(indexProperty);
         const item = this._items[index];
         const exemptKey = item ? this.getKey(item, Number(index)) : undefined;
         const newIndex = exemptKey ? newKeyToIndexMap[exemptKey] : index;
 
         if (newIndex === undefined) {
-          // We don't know the index of the item any more so it's either moved or removed.
-          // In this case we reset the entire selection.
-          this.setAllSelected(false);
-          break;
+          // The item has likely been replaced or removed.
+          hasSelectionChanged = true;
         } else {
           // We know the new index of the item. update the existing exemption table.
           newExemptedIndicies[newIndex] = true;
+          newExemptedCount++;
           hasSelectionChanged = hasSelectionChanged || newIndex !== index;
         }
       }
     }
 
+    if (this._items && this._exemptedCount === 0 && items.length !== this._items.length && this._isAllSelected) {
```

Comment:
should there be a test case covering this?

Reply:
Done.